ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -646,24 +679,37 @@ public String getResourceNameForPartition(AmbryPartition partition) {
 
     /**
      * {@inheritDoc}
-     * If dcName is null, then get replicas by given state from all datacenters.
+     * If dcOrClusterName is null, then get replicas by given state from all datacenters.
      * If no routing table snapshot is found for dc name, or no resource name found for given partition, return empty list.
      */
     @Override
     public List<AmbryReplica> getReplicaIdsByState(AmbryPartition partition, ReplicaState state, String dcName) {
       List<AmbryReplica> replicas = new ArrayList<>();
-      for (DcInfo dcInfo : dcToDcInfo.values()) {
-        String dc = dcInfo.dcName;
-        if (dcName == null || dcName.equals(dc)) {
-          String resourceName = partitionToResourceNameByDc.get(dc).get(partition.toPathString());
-          dcToRoutingTableSnapshotRef.get(dcInfo.dcName)
-              .get()
-              .getInstancesForResource(resourceName, partition.toPathString(), state.name())
-              .stream()
-              .map(instanceConfig -> instanceNameToAmbryDataNode.get(instanceConfig.getInstanceName()))
-              .map(dataNode -> ambryDataNodeToAmbryReplicas.get(dataNode).get(partition.toPathString()))
-              .filter(Objects::nonNull)
-              .forEach(replicas::add);
+      if (clusterMapConfig.clusterMapUseAggregatedView) {
+        String resourceName = dcName != null ? partitionToResourceNameByDc.get(dcName).get(partition.toPathString())
+            : getResourceNameForPartition(partition);
+        globalRoutingTableSnapshotRef.get()
+            .getInstancesForResource(resourceName, partition.toPathString(), state.name())
+            .stream()
+            .map(instanceConfig -> instanceNameToAmbryDataNode.get(instanceConfig.getInstanceName()))
+            .filter(dataNode -> dcName == null || dataNode.getDatacenterName().equals(dcName))
+            .map(dataNode -> ambryDataNodeToAmbryReplicas.get(dataNode).get(partition.toPathString()))
+            .filter(Objects::nonNull)
+            .forEach(replicas::add);
+      } else {
+        for (DcInfo dcInfo : dcToDcInfo.values()) {
+          String dc = dcInfo.dcName;
+          if (dcName == null || dcName.equals(dc)) {
+            String resourceName = partitionToResourceNameByDc.get(dc).get(partition.toPathString());
+            dcToRoutingTableSnapshotRef.get(dcInfo.dcName)
+                .get()
+                .getInstancesForResource(resourceName, partition.toPathString(), state.name())
+                .stream()
+                .map(instanceConfig -> instanceNameToAmbryDataNode.get(instanceConfig.getInstanceName()))
+                .map(dataNode -> ambryDataNodeToAmbryReplicas.get(dataNode).get(partition.toPathString()))
+                .filter(Objects::nonNull)
+                .forEach(replicas::add);
+          }
```

Comment:
should probably merge these two method. 
This method can be
```
Map<ReplicaState, List<AmbryReplica>> result = new HashMap<>();
getReplicaIdByStates(result, partition, Collections.singleton(state), dcName);
return result.get(state);
```

Reply:
+1 in combining.  The method below can for(ReplicateState : states) call above method.